Add Orthogonal Subspace Fine-Tuning (OSF) Tuner for Parameter-Efficient Continual Learning

NikhilNayak-debug opened the pull request:
**Summary**

This PR adds a new parameter-efficient fine-tuning method called **Orthogonal Subspace Fine-Tuning (OSF)** to the PEFT library. OSF enables continual learning in LLMs by freezing the high-rank subspace of weight matrices and fine-tuning only the low-rank directions. This approach constrains updates to be orthogonal to previously important directions, thereby mitigating catastrophic forgetting without increasing parameter count.

---

**Issue for this PR on PEFT repository**

Tracked in [PEFT Issue #2648](https://github.com/huggingface/peft/issues/2648)

---

**Key Features**

* Implements a new `OSFConfig`, `OSFModel`, and tuner class under `src/peft/tuners/osf/` following PEFT's standard API
* Integrates seamlessly with the `get_peft_model` API:

  ```python
  from peft import OSFConfig, get_peft_model
  peft_model = get_peft_model(base_model, OSFConfig(target_modules=[...]))
  ```
* Adds utility functions for:

  * Weight matrix decomposition using SVD
  * Gradient projection onto the low-rank subspace via backward gradient hooks
* Automatically enforces orthogonality constraints during training without requiring optimizer wrapping
* Will include tests for saving, loading, and applying the OSF adapter in `tests/test_custom_models.py`
* Exports relevant modules at the package level for easier use with other PEFT components

---

**Notes**

* The current implementation does not include layerwise importance-based rank estimation (e.g., cosine similarity of inputs and activations), but can be added in future iterations
* Merging/unmerging is not supported, as the original weights are decomposed and modified in-place
* Compared to LoRA, OSF performs a constrained update over the original weight matrix without introducing new trainable parameters, maintaining exact model architecture post-training

---

**Background**

**This implementation is based on the method described in our paper:**
***Sculpting Subspaces: Constrained Full Fine-Tuning in LLMs for Continual Learning***
[Paper on arXiv](https://arxiv.org/abs/2504.07097) · [Project Repository](https://github.com/NikhilNayak-debug/mini_trainer)

---

githubnemo reviewed (commented):
Nice! Thanks for the thorough update, that's a good step forward.
A minor nit: Several files are missing the copyright notice, please make sure to include them in new source files (also make sure that they are not outdated, i.e. include the current year).

I like that you already implemented several (custom) tests, I think that's super helpful. Let's also add some tests to `test_decoder_models.py` and `test_encoder_decoder_models.py` similar to the test in `test_custom_models.py` when you think the implementation can move forward in testing. Let's move the skips for convolutions to `testing_common.py`, there are already similar exceptions in place.
Two bigger topics:

1. `ModelWithOSF` seems to re-invent PEFT functionality inside PEFT, specifically the layer targeting + replacement portion. Let's streamline OSF with other tuners, i.e. have implementations for specific layers and by implementing `inject_adapter`, `_create_new_module` and `_create_and_replace` to make it easier to branch out to other layer types / quantizations. The LoRA implementation maybe helpful, e.g. `peft.tuners.lora.layers.LoraLayer` contains specific layers for `Linear` and `Conv*d` specifics (no need to implement Conv now, of course). I can see that this conflicts with using a dict for specifying the top-k ranks per module. How about using `target_modules` and a singular value for the topk rank (e.g., `config.topk_r`) which can default to `None` (-> uses 50% of min(shape)). Every targeted module gets that topk rank or an automatic 50% one. We could also add something like `rank_pattern` from LoRA to define exceptions (see `lora.model.py` -> `_create_and_replace`). WDYT?
Example config:
```python
OSFConfig(
  target_modules='all-linear',
  topk_r=None,
  rank_pattern={
    'q_proj': 10,
  }
)
```

2. It's not possible to use more than one adapter of OSF since the base model is modified and we therefore cannot switch between adapters (could be handy in pipeline scenarios where one model is used at several places with different adapters, for example). I left a comment at `decompose_weight_matrix` to discuss this.

Once we're done with the general implementation I think it'd be super if we could add an experiment to the MetaMathQA comparison suite so that we can compare OSF directly to other implementations.

NikhilNayak-debug commented on src/peft/__init__.py:
```diff
@@ -212,6 +226,8 @@
     "load_peft_weights",
     "prepare_model_for_kbit_training",
+    "project_gradient_to_orthogonal_space",
+    "reconstruct_weight_matrix",
```
Done removed.

NikhilNayak-debug commented on src/peft/utils/osf_utils.py:
```diff
@@ -0,0 +1,306 @@
+    config: dict[str, int] = {}
+    for name, param in model.named_parameters():
+        if any(pat in name for pat in target_patterns) and len(param.shape) == 2:
+            top_k = int(math.floor(min(param.shape) * 0.5))
```
Thank you this makes it cleaner, changed it!

NikhilNayak-debug commented on src/peft/utils/osf_utils.py:
```diff
@@ -0,0 +1,297 @@
+        if hasattr(dU, "_local_tensor"):
+            dU._local_tensor.copy_(local_dU)
+        else:
+            dU.copy_(local_dU)
```
You are right! Removed the unnecessary DTensor distinction. Simplified to just use `dU.copy_(local_dU)` and `dV.copy_(local_dV)`.

NikhilNayak-debug commented on src/peft/utils/osf_utils.py:
```diff
@@ -0,0 +1,297 @@
+
+        def hook(grad, svd=svd_dict):
+            project_gradient_to_orthogonal_space(svd)
+            return grad
```
Done!

NikhilNayak-debug commented on src/peft/utils/osf_utils.py:
```diff
@@ -0,0 +1,297 @@
+
+        module_svd.U_low.register_hook(hook)
+        module_svd.S_low.register_hook(hook)
+        module_svd.V_low.register_hook(hook)
```
Done. Added proper hook handle management:
  - `attach_gradient_hooks()` now stores all hook handles in `model.hook_handles`
  - Added `detach_gradient_hooks()` function to properly remove hooks and prevent memory leaks
  - Updated `reinitialize_svd()` to clean up existing hooks before reattaching new ones

NikhilNayak-debug commented:
> 1. `ModelWithOSF` seems to re-invent PEFT functionality inside PEFT, specifically the layer targeting + replacement portion. Let's streamline OSF with other tuners, i.e. have implementations for specific layers and by implementing `inject_adapter`, `_create_new_module` and `_create_and_replace` to make it easier to branch out to other layer types / quantizations. The LoRA implementation maybe helpful, e.g. `peft.tuners.lora.layers.LoraLayer` contains specific layers for `Linear` and `Conv*d` specifics (no need to implement Conv now, of course). I can see that this conflicts with using a dict for specifying the top-k ranks per module. How about using `target_modules` and a singular value for the topk rank (e.g., `config.topk_r`) which can default to `None` (-> uses 50% of min(shape)). Every targeted module gets that topk rank or an automatic 50% one. We could also add something like `rank_pattern` from LoRA to define exceptions (see `lora.model.py` -> `_create_and_replace`). WDYT?
>    Example config:
> 
> ```python
> OSFConfig(
>   target_modules='all-linear',
>   topk_r=None,
>   rank_pattern={
>     'q_proj': 10,
>   }
> )
> ```
@githubnemo great suggestion in response to the first bigger topic raised I have implemented the minimal PEFT integration changes:

  **What we implemented:**
  - ✅ OSF layer classes (`OSFLayer`, `Linear`) similar to LoRA's structure
  - ✅ `_create_and_replace` method for proper layer replacement following PEFT patterns
  - ✅ Updated config to use `target_modules` and `effective_rank` (renamed from `topk_r`)
  - ✅ Added `rank_pattern` support for per-module rank exceptions, just like LoRA

  **Scope decisions we made:**
  - **Only implemented `_create_and_replace`** (not `inject_adapter` or `_create_new_module`) since OSF's use case only requires layer replacement as of now
  - **Kept existing functionality intact** - all SVD decomposition, gradient projection, and hook management preserved as is

  **Key files changed:**
  - `src/peft/tuners/osf/layer.py` - New OSF layer classes
  - `src/peft/tuners/osf/model.py` - Added `_create_and_replace` method
  - `src/peft/tuners/osf/config.py` - Updated config format
  - `src/peft/utils/constants.py` - Added `TRANSFORMERS_MODELS_TO_OSF_TARGET_MODULES_MAPPING`

These changes integrate the OSF method modularly into PEFT.

NikhilNayak-debug commented on examples/orthogonal_subspace_learning/README.md:
````diff
@@ -0,0 +1,21 @@
+loss.backward()
+optimizer.step()
+optimizer.zero_grad()
+```
````
@githubnemo done! Moved documentation to `docs/source/package_reference/osf.md` with paper link, abstract, description, autodoc references, and training advice.

Also added placeholder in `examples/orthogonal_subspace_learning/README.md` for the runnable example, will implement a comprehensive example with multiple tasks after the current implementation is complete.

githubnemo reviewed (commented):
Thanks for the detailed feedback and your changes.

I think that the re-structuring of `OSFModel` is almost complete and most of the comments are rather minor. As far as I can see the adhoc `ModelWithOSF` is replaced by `OSFModel` and `OSFLayer` and can be removed - good progress!

I think this is a good time remove outdated code, to merge with `main`, run `make style` and run the tests to see if there's still something going horribly wrong.

Let's discuss whether we want to implement the importance score now or leave it up for implementation later. If I'm not mistaken I think that the importance score can technically be added later since it would compute the effective rank of layers based on two new hyper-parameters, so in that sense it is modular. Since it is quite a crucial part of the paper and is touted to improve multi-task learning (arguably one of the big selling points of OSF) I wonder if it should be included from the get-go. What's your opinion on that?

Regardless, I think we can a MetaMathQA experiment rather soon and check if there are major problems with memory consumption or runtime.

NikhilNayak-debug commented on docs/source/package_reference/osf.md:
```diff
@@ -0,0 +1,240 @@
+
+# OSF (Orthogonal Subspace Fine-tuning)
+
+Orthogonal Subspace Fine-tuning ([OSF](https://arxiv.org/abs/2504.07097)) is a PEFT method designed for continual learning that constrains parameter updates to be orthogonal to previously important directions. This approach enables full fine-tuning while preventing catastrophic forgetting without requiring additional parameters or storing previous gradients.
```
Done!

NikhilNayak-debug commented on docs/source/package_reference/osf.md:
```diff
@@ -0,0 +1,240 @@
+
+### Best Practices
+
+1. **Effective Rank Selection**: Start with `effective_rank=None` (automatic 50% rank) and adjust based on task complexity
```
Fixed!

NikhilNayak-debug commented on tests/testing_common.py:
```diff
@@ -234,6 +235,15 @@ def test_something(model_id, config_kwargs):
         raise
 
 
+def _skip_if_merging_not_supported(model_id, config_cls):
+    """Skip tests for Conv2dGroups models or OSF configs where merging is not supported."""
+    if model_id in ["Conv2dGroups", "Conv2dGroups2"] or issubclass(config_cls, OSFConfig):
+        pytest.skip(
+            f"Skipping test for {model_id} with {config_cls} as merging is not supported. "
+            "(See https://github.com/huggingface/peft/pull/2403 for details)"
+        )
+
```
Makes sense, done!

NikhilNayak-debug commented on tests/test_osf_utils.py:
Done!

NikhilNayak-debug commented on tests/test_osf_utils.py:
```diff
@@ -0,0 +1,118 @@
+    )
+
+
+def test_osf_config_roundtrip():
+    cfg = OSFConfig(target_svd_config={"linear.weight": 2})
+    with TemporaryDirectory() as tmp:
+        cfg.save_pretrained(tmp)
+        loaded = OSFConfig.from_pretrained(tmp)
+    assert cfg.target_svd_config == loaded.target_svd_config
```
Yup removed!

NikhilNayak-debug commented on tests/test_osf_utils.py:
```diff
@@ -0,0 +1,118 @@
+    assert cfg.target_svd_config == loaded.target_svd_config
+
+
+def test_osf_save_and_load_model():
+    torch.manual_seed(0)
+    model = DummyModel(DummyConfig())
+    cfg = OSFConfig(target_svd_config={"linear.weight": 2})
+    wrapped = get_peft_model(model, cfg)
+    x = torch.randn(2, 8)
+    y_ref = wrapped(x)
+    with TemporaryDirectory() as tmp:
+        wrapped.save_pretrained(tmp)
+        torch.manual_seed(0)
+        base = DummyModel(DummyConfig())
+        loaded = PeftModel.from_pretrained(base, tmp)
+        y = loaded(x)
+    assert_close(y, y_ref, atol=1e-5, rtol=1e-5)
```
Makes sense this save/load test is redundant.

NikhilNayak-debug commented on tests/test_osf_utils.py:
```diff
@@ -0,0 +1,118 @@
+        return self.linear(x)
+
+
+def test_osf_get_peft_model_preserves_output():
+    torch.manual_seed(0)
+    model = DummyModel(DummyConfig())
+    x = torch.randn(2, 8)
+    y_ref = model(x)
+    wrapped = get_peft_model(model, OSFConfig())
+    y = wrapped(x)
+    assert_close(y, y_ref, atol=1e-5, rtol=1e-5)
```
Deleted test_osf_get_peft_model_preserves_output(), since generic adapter tests already verify that wrapping with a config does not change outputs.

NikhilNayak-debug commented on src/peft/tuners/osf/model.py:
```diff
@@ -0,0 +1,106 @@
+    def _prepare_adapter_config(self, peft_config, model_config):
+        return peft_config
+
+    def inject_adapter(self, model: nn.Module, adapter_name: str, autocast_adapter_dtype: bool = True, low_cpu_mem_usage: bool = False) -> None:
+        # For now, keep using the legacy approach
+        # TODO: Refactor to use _create_and_replace pattern
+        svd_cfg = auto_generate_target_osf_config(model)
+            
+        OSFCls = create_osf_model_class(model.__class__)
+        base_cfg = getattr(model, "config", None)
+        osf_model = OSFCls(base_cfg, svd_config=svd_cfg, initialize_svd=False)
+        osf_model.load_state_dict(model.state_dict())
+        osf_model.reinitialize_svd()
+        attach_gradient_hooks(osf_model)
+        self.model = osf_model
```
Thanks implemented this refactor.

  - Removed the custom inject_adapter we now use BaseTuner.inject_adapter for target discovery.
  - Added OSFModel._check_target_module_exists delegating to tuners_utils.check_target_module_exists.
  - Changed _create_and_replace method to operate in-place: update existing OSFLayer or replace via dispatch_default and _replace_module.

NikhilNayak-debug commented on src/peft/tuners/osf/layer.py:
```diff
@@ -0,0 +1,274 @@
+        self._osf_V_high[adapter_name] = svd_dict["V_high"]
+        
+        # Create module for trainable low-rank components
+        svd_module = nn.Module()
```
Thank you for the suggestion. ParameterDict makes it explicit these are trainable parameters and improves the printed structure (no anonymous Module()). Now using ParameterDict for OSF's low‑rank components and updated usages accordingly.

NikhilNayak-debug commented on src/peft/tuners/osf/utils.py:
```diff
@@ -0,0 +1,337 @@
+        dV.copy_(local_dV)
+
+
+def attach_gradient_hooks(model: nn.Module) -> None:
+    """Attach gradient hooks to project OSF gradients automatically."""
+    if not hasattr(model, "svd_params"):
+        return
+    
+    # Initialize hook_handles if not exists
+    if not hasattr(model, "hook_handles"):
+        model.hook_handles = []
+    
+    for safe_name, module_svd in model.svd_params.items():
+        svd_dict = {
+            "U_high": getattr(model, f"{safe_name}_U_high"),
+            "S_high": getattr(model, f"{safe_name}_S_high"),
+            "V_high": getattr(model, f"{safe_name}_V_high"),
+            "U_low": module_svd.U_low,
+            "S_low": module_svd.S_low,
+            "V_low": module_svd.V_low,
+        }
+
+        def hook(grad):
+            project_gradient_to_orthogonal_space(svd_dict)
+            return grad
+
+        # Store hook handles for later cleanup
+        handle_u = module_svd.U_low.register_hook(hook)
+        handle_s = module_svd.S_low.register_hook(hook)
+        handle_v = module_svd.V_low.register_hook(hook)
+        
+        model.hook_handles.extend([handle_u, handle_s, handle_v])
+
+
+def detach_gradient_hooks(model: nn.Module) -> None:
+    """Remove all gradient hooks that were attached by attach_gradient_hooks."""
+    if hasattr(model, "hook_handles"):
+        for handle in model.hook_handles:
+            handle.remove()
+        model.hook_handles.clear()
```
Good point, switched to layer-local hooks and removed the model-level helpers.

NikhilNayak-debug commented on src/peft/tuners/osf/layer.py:
```diff
@@ -0,0 +1,274 @@
+            warnings.warn("Already unmerged. Nothing to do.")
+            return
+            
+        # For OSF, unmerging means restoring the original weight
+        # Since we modify the weight in-place, we need to store the original weight
+        # This is a limitation of the current OSF implementation
+        warnings.warn("OSF does not support unmerging. Original weights are permanently modified.")
```
Thank you for the suggestion. For OSF unmerge would restore the original weights from which the new training run was started. Storing originals per target would add memory and complexity but still would not give true step-wise reversibility. For now shall we keep merge/unmerge as not implemented?

githubnemo replied:
You're right, it doesn't make a lot of sense.

NikhilNayak-debug commented on src/peft/tuners/osf/model.py:
```diff
@@ -0,0 +1,106 @@
+    def __init__(self, model, config, adapter_name, low_cpu_mem_usage: bool = False):
+        super().__init__(model, config, adapter_name, low_cpu_mem_usage=low_cpu_mem_usage)
+
+    def _prepare_adapter_config(self, peft_config, model_config):
+        return peft_config
```
Done implemented exactly that.

Added _prepare_adapter_config mirroring LoRA:
- Infers defaults from TRANSFORMERS_MODELS_TO_OSF_TARGET_MODULES_MAPPING when target_modules is None.
- Raises a clear ValueError if model_type is not mapped.

NikhilNayak-debug commented on src/peft/tuners/osf/model.py:
```diff
@@ -0,0 +1,106 @@
+    def set_adapter(self, adapter_name):
+        self.active_adapter = adapter_name
+
+    def unload(self):
+        raise NotImplementedError("OSF models cannot be unloaded yet")
+
+    def merge_adapter(self, *args, **kwargs):
+        raise NotImplementedError("OSF models do not support merging")
+
+    def unmerge_adapter(self, *args, **kwargs):
+        raise NotImplementedError("OSF models do not support merging")
+
+    def merge_and_unload(self, *args, **kwargs):
+        raise NotImplementedError("OSF models do not support merging")
```
If we are keeping OSF non-mergeable for now, no code change is required here.

NikhilNayak-debug commented on src/peft/tuners/osf/layer.py:
```diff
@@ -0,0 +1,274 @@
+
+    def __init__(self, base_layer: nn.Module, **kwargs) -> None:
+        self.base_layer = base_layer
+        self.effective_rank = {}
```
Yes that is exactly right and effective_rank is also more conceptually descriptive here for what we are trying to do with OSF in identifying the important subspace (effective rank of the matrix).

NikhilNayak-debug commented:
@githubnemo added MetaMathQA experiment results. OSF achieves the highest accuracy at 55.72% among all PEFT methods in the benchmark! 😊

  Top results for comparison:
  - osf--llama-3.2-3B-default: 0.5572
  - lora--llama-3.2-3B-rank64-rslora: 0.5299
  - bone--llama-3.2-3B-bat: 0.5171
  - bone--llama-3.2-3B-default: 0.5080
  - randlora--llama-3.2-3B-default: 0.5072

  Memory consumption and runtime look okay thus far as well.

NikhilNayak-debug committed:
Orthogonal Subspace Learning: changes for the OSF method

rebasing to make use of simplified basetuner implementation and adding more experiment results

fixing style, quality, etc in the code

Make style

fixing CI and other test cases

githubnemo commented on src/peft/tuners/osf/utils.py:
```diff
@@ -0,0 +1,116 @@
+    U, S, Vt = torch.linalg.svd(W, full_matrices=False)
+    k = min(top_k, S.shape[0])
+
+    svd = {
+        "U_high": U[:, :k].contiguous().detach().to(device=device_local, dtype=orig_dtype),
+        "S_high": S[:k].contiguous().detach().to(device=device_local, dtype=orig_dtype),
+        "V_high": Vt[:k, :].contiguous().detach().to(device=device_local, dtype=orig_dtype),
+        "U_low": nn.Parameter(U[:, k:].contiguous().detach().to(device=device_local, dtype=orig_dtype)),
+        "S_low": nn.Parameter(S[k:].contiguous().detach().to(device=device_local, dtype=orig_dtype)),
+        "V_low": nn.Parameter(Vt[k:, :].contiguous().detach().to(device=device_local, dtype=orig_dtype)),
+        "rank_high": k,
+    }
```
Maybe I'm misunderstanding something but I understand `effective_rank` to be the best guess estimate of the effective rank of the weight matrix - the rest is assumed 'free' for us to modify (thus, the "low" portion has `S.shape[0] - effective_rank` space for the trained task).

If that is the case, then the example in the documentation for continual learning is wrong as it decreases the effective rank over time (therefore growing the low part, shrinking the high part).

In case the above is correct the effective rank parameter should be explained better in the config to avoid mistakes. Maybe it is even worth thinking about swapping the semantics to be more akin to LoRA's `r` since I can imagine a few people being surprised by using `effective_rank=1` and running into OOM errors. It might also be easier to start from the bottom up and taking more space than guessing a high number and decreasing it. WDYT?

githubnemo replied:
Thanks. I'm OK with keeping the effective rank semantic as is but since the default is `0.5 * min(dim)` and we're therefore supporting fractions I, as a user, would expect fractions to work in general (e.g., `effective_rank=0.8`, translating to `0.8 * min(dim)`. I think that would also help in cases where I want to increment the effective rank over time since I can do so in relative steps.

NikhilNayak-debug replied:
@githubnemo this is a good suggestion. I have made the change to now accept `effective_rank` as both an integer representing the actual rank as well as a ratio between 0 to 1 as a floating number to represent the percentage of the smaller dimension to retain.

githubnemo commented on src/peft/tuners/osf/model.py:
```diff
@@ -0,0 +1,192 @@
+    def unload(self):
+        raise NotImplementedError("OSF models cannot be unloaded yet")
+
+    def merge_adapter(self, *args, **kwargs):
+        raise NotImplementedError("OSF models do not support merging")
+
+    def unmerge_adapter(self, *args, **kwargs):
+        raise NotImplementedError("OSF models do not support merging")
+
+    def merge_and_unload(self, *args, **kwargs):
+        raise NotImplementedError("OSF models do not support merging")
```
```suggestion
    def unmerge_adapter(self, *args, **kwargs):
        raise NotImplementedError("OSF models do not support merging")
```

IIUC only `unmerge` is unsupported. Since #2771 is now merged we can remove these implementations and use the `BaseTuner` implementation which calls `layer.merge` under the hood.

NikhilNayak-debug replied:
This makes sense made the following changes:

- Removed OSF specific overrides for merge_adapter and merge_and_unload to use BaseTuner's implementations.
- Kept unmerge unsupported to raise NotImplementedError.
- merge and merge_and_unload work via BaseTuner and unmerge remains explicitly disallowed for OSF.

githubnemo reviewed (commented):
Sorry for the late reply, I was at a conference.

The changes look very good! There was quite a large PR merged in the mean time that refactored a good portion of the `BaseTuner` infrastructure (#2771) which means that you need a lot less code now - I hope I highlighted all occurrences.

I'm currently in the process of reproducing the MetaMathQA results you posted. One thing I noticed is that there are more layers targeted and the default effective rank (`min(shape) // 2`) is used which is using way more parameters than other methods. While it is certainly good to see that OSF is better than full fine-tuning it would be a fairer comparison to match the trainable parameter counts of the other methods.

NikhilNayak-debug commented on docs/source/package_reference/osf.md:
```diff
@@ -0,0 +1,236 @@
+train_task(model, task_2_data)
+
+# Task 3: recompute again and expand preserved subspace further
+base_model = model.base_model.model
```
I removed the method implementation that raises the NotImplementedError. BaseTuner.unload() already handles removing adapter wrappers and restoring the underlying base layers. OSF's layers inherit BaseTunerLayer so this works out of the box.

githubnemo commented:
Thanks for the updated experiments and the fixes!

I think this can be merged after these few nits are fixed. It would be good to have a runnable example but if you want we can merge this first and do the example in a separate PR, whichever suits you best. It could also be a chance to showcase the continual learning benefits of OSF to contrast the weaker parameter efficiency.

Before merging, let's remove the MetaMathQA results since we're running them on our runner after merging anyway. 
Let's also remove the big MetaMathQA experiment since it won't run on our runner and is not comparable to the other experiments and would, therefore, rot far away from others as a lone outlier.

Don't forget to run `make style` so that we can run the CI and check if there are any open issues from there.

githubnemo commented on docs/source/_toctree.yml:
```diff
@@ -143,6 +145,8 @@
     - local: package_reference/helpers
       title: Helpers
+    - local: package_reference/osf_utils
+      title: OSF utilities
```
This is not relevant anymore, right?

NikhilNayak-debug replied:
Yes removed!

githubnemo reviewed (commented):
Nice :) Only a few comments left.

Please check the CI as there seem to be some errors left.

NikhilNayak-debug commented:
@githubnemo this is ready for review. There are 36 test cases failing, but they are also failing in the upstream main branch so they are not related to the OSF changes. All CI related issues should be addressed now.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_2685). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

NikhilNayak-debug committed:
fixing couple of test errors

NikhilNayak-debug commented:
@githubnemo could you start the workflow again, fixed a couple of remaining test case errors!

githubnemo commented:
Sure thing. You can also run this using `pytest tests`, you can also select for OSF using `pytest -k osf tests`

githubnemo reviewed (approved):
Thank you! This looks good to me :)

Let's merge this and work on the example in a new PR. I think this is a nice addition, thanks again for your continued efforts!

githubnemo commented:
@NikhilNayak-debug just a heads-up, we recently fixed a bug in the MetaMathQA benchmark regarding forgetting (#3419) and now OSF is on the pareto frontier of forgetting vs. accuracy (https://huggingface.co/spaces/peft-internal-testing/PEFT-method-comparison).
